Refactor get numerical jacobian

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54480 Implement faster gradcheck but run both fast and slow
* #54479 Factor out numerical logic
* #54378 Refactor get numerical jacobian to calculate wrt all outputs at once
* **#54092 Refactor get numerical jacobian**

This is the first of several refactors to get numerical jacobian:
This one just moves some logic around as to try to split the get_numerical_jacobian function into smaller more manageable functions:
- compute_gradient is now no longer nested, but we have to pass in the parameters instead
- iter_tensor extracts out the logic of iterating through different types of tensors (the code should be almost the exact same here except for instead of calling into the update jacobian function, we yield the arguments instead)

Differential Revision: [D27354268](https://our.internmc.facebook.com/intern/diff/D27354268)

soulitzer committed:
Implement fast gradcheck

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Implement fast gradcheck"


- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True
- For individual tests that are failing, override with fast_mode=False while we investigate

TODO:
- sparse, mkldnn
- 

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True
- For individual tests that are failing, override with fast_mode=False while we investigate

TODO:
- sparse, mkldnn
- 

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True
- For individual tests that are failing, override with fast_mode=False while we investigate

TODO:
- sparse, mkldnn
- 

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True
- For individual tests that are failing, override with fast_mode=False while we investigate

TODO:
- sparse, mkldnn
- 

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True
- For individual tests that are failing, override with fast_mode=False while we investigate

TODO:
- sparse, mkldnn
- 

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True
- For individual tests that are failing, override with fast_mode=False while we investigate

TODO:
- sparse, mkldnn
- 

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True
- For individual tests that are failing, override with fast_mode=False while we investigate

TODO:
- sparse, mkldnn
- 

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True
- For individual tests that are failing, override with fast_mode=False while we investigate

TODO:
- sparse, mkldnn
- 

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. 

Other 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- Factors out code that iterates through different types of tensors.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. 

Other 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- Factors out code that iterates through different types of tensors.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. 

Other 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- Factors out code that iterates through different types of tensors.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get numerical jacobian"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"


This is the first of several refactors to get numerical jacobian:
This one just moves some logic around as to try to split the get_numerical_jacobian function into smaller more manageable functions:
- compute_gradient is now no longer nested, but we have to pass in the parameters instead
- iter_tensor extracts out the logic of iterating through different types of tensors (the code should be almost the exact same here except for instead of calling into the update jacobian function, we yield the arguments instead)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"


This is the first of several refactors to get numerical jacobian:
This one just moves some logic around as to try to split the get_numerical_jacobian function into smaller more manageable functions:
- compute_gradient is now no longer nested, but we have to pass in the parameters instead
- iter_tensor extracts out the logic of iterating through different types of tensors (the code should be almost the exact same here except for instead of calling into the update jacobian function, we yield the arguments instead)

[ghstack-poisoned]

soulitzer commented:
Add checks similar to what we have for analytical jacobian to verify shape/type.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -47,7 +47,77 @@ def iter_tensors(x: Union[torch.Tensor, Iterable[torch.Tensor]], only_requiring_
 
-def get_numerical_jacobian(fn, input, target=None, eps=1e-3, grad_out=1.0):
+def iter_tensor(x_tensor):
+    # Enumerates over a tensor and provides a corresponding flat index that translates
+    # to a given rol/col in the jacobian matrix. iter_tensor also returns a
```
I've never read the get_numeric_jacobian before so this is a bit magical to me. How does the "jacobian matrix" in the example correspond to the "tensor"? What is an example of an iterator that this function outputs?

soulitzer replied:
For strided, `d_idx` corresponds to the correct column in the jacobian matrix. And the order is the same as as if we flatten a contiguous tensor. An example is that for a tensor t with size (2, 2): it will yield `x, 0, (0, 0)`, `x, 1, (0, 1)`, `x, 2, (1, 0)`, `x, 3, (1, 1)` where x is the `t.data` of the original tensor. The input t has numel 4, so Jacobian should have 4 columns. So having a `d_idx` of 3 and `idx` of `(1, 1)` indicates that perturbing `t[(1, 1)]` allows us to updating the third (last) column of any jacobian corresponding to this particular input.

soulitzer replied:
Mkldnn is similar, but requires us to do a dense copy first. Sparse is a bit more complicated since we need to reverse engineer the strides from the size, and also consider that it may have both sparse/dense dims.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -67,97 +137,31 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3, grad_out=1.0):
+    for x_tensor, d in zip(x_tensors, j_tensors):
+        is_mkldnn = x_tensor.layout == torch._mkldnn  # type: ignore # no attr _mkldnn
+        for x, idx, d_idx in iter_tensor(x_tensor):
+            # compute_jacobian only works for pure real or pure imaginary delta
```
What "compute_jacobian" is this referring to?

soulitzer replied:
Hmm, that comment should probably be updated to "computing the jacobian". It should refer to computing the jacobian in general, though the reason why I don't quite understand yet. It might be somewhat related to the fact that we are trying to compute the derivatives of non-analytic functions.

zou3519 reviewed (approved):
I'm confused on what exactly iter_tensor produces (maybe an example in the doc string would clear it up), but this PR is a straightforward refactor and LGTM

soulitzer committed:
Update on "Refactor get numerical jacobian"


This is the first of several refactors to get numerical jacobian:
This one just moves some logic around as to try to split the get_numerical_jacobian function into smaller more manageable functions:
- compute_gradient is now no longer nested, but we have to pass in the parameters instead
- iter_tensor extracts out the logic of iterating through different types of tensors (the code should be almost the exact same here except for instead of calling into the update jacobian function, we yield the arguments instead)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"


This is the first of several refactors to get numerical jacobian:
This one just moves some logic around as to try to split the get_numerical_jacobian function into smaller more manageable functions:
- compute_gradient is now no longer nested, but we have to pass in the parameters instead
- iter_tensor extracts out the logic of iterating through different types of tensors (the code should be almost the exact same here except for instead of calling into the update jacobian function, we yield the arguments instead)

Differential Revision: [D27354268](https://our.internmc.facebook.com/intern/diff/D27354268)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"


This is the first of several refactors to get numerical jacobian:
This one just moves some logic around as to try to split the get_numerical_jacobian function into smaller more manageable functions:
- compute_gradient is now no longer nested, but we have to pass in the parameters instead
- iter_tensor extracts out the logic of iterating through different types of tensors (the code should be almost the exact same here except for instead of calling into the update jacobian function, we yield the arguments instead)

Differential Revision: [D27354268](https://our.internmc.facebook.com/intern/diff/D27354268)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian"


This is the first of several refactors to get numerical jacobian:
This one just moves some logic around as to try to split the get_numerical_jacobian function into smaller more manageable functions:
- compute_gradient is now no longer nested, but we have to pass in the parameters instead
- iter_tensor extracts out the logic of iterating through different types of tensors (the code should be almost the exact same here except for instead of calling into the update jacobian function, we yield the arguments instead)

Differential Revision: [D27354268](https://our.internmc.facebook.com/intern/diff/D27354268)

[ghstack-poisoned]
